### PR TITLE
Add storage layer integration

### LINKED ISSUE #30: As a user, I want generated diagrams to correspond to my description with high accuracy
## Problem

v0.0.1 of the product relies on a general codex model which performs poorly on non-trivial prompts. We [identified](https://github.com/kislerdm/diagramastext/discussions/23) the strategy to improve the app's output quality. It requires use of the prompts submitted by our users.

## Proposed Solution

Store the input prompt and the model generated graph to a database.

## Acceptance Criteria
- Every prompt and model-generated graph is stored to a Neon postgres db

### PR DESCRIPTION
**Pull request by kislerdm**
## Why do we need it

To improve diagram generation quality

## What changed

- Added postgres client
- Add integration logic to preserve user's input and model's prediction result
- Adjusted CI jobs

Closes #30

### CODE REVIEW

File: core/cmd/cli/main.go [hunk 1]
```diff
@@ -1,3 +1,6 @@
+//go:build !unittest
```
Comment (diegohordi) [inline core/cmd/cli/main.go:1]: **question** WDYT https://peter.bourgon.org/blog/2021/04/02/dont-use-build-tags-for-integration-tests.html?

File: core/cmd/lambda/handler_test.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 	"time"
 
 	"github.com/aws/aws-lambda-go/events"
+	"github.com/aws/aws-lambda-go/lambdacontext"
 	"github.com/kislerdm/diagramastext/core"
 	coreHandler "github.com/kislerdm/diagramastext/core/handler"
```
Comment (diegohordi) [inline core/cmd/lambda/handler_test.go:15]: **question** Why not rename the `handler` package to `handlers` so we don't need to give an import name that also might collide with some variable name?

File: core/cmd/lambda/main.go [hunk 3]
```diff
@@ -67,14 +76,14 @@ func (h corsHeaders) setHeaders(resp events.APIGatewayProxyResponse) events.APIG
...
+	clientStorage core.ClientStorage,
 ) func(
 	ctx context.Context, req events.APIGatewayProxyRequest,
 ) (events.APIGatewayProxyResponse, error) {
 	return func(
 		ctx context.Context, req events.APIGatewayProxyRequest,
 	) (events.APIGatewayProxyResponse, error) {
-		// FIXME: add proper sink to preserve user's requests for model fine-tuning
-		log.Println(req.Body)
+		defer func() { _ = clientStorage.Close(ctx) }()
```
Comment (diegohordi) [inline core/cmd/lambda/main.go:86]: **question** why don't we log the error instead of silent it?

File: core/cmd/lambda/main.go [hunk 4]
```diff
@@ -95,11 +104,40 @@ func handler(
 			), err
 		}
 
+		userID := readUserID(req.Headers)
+		requestID := readRequestID(ctx)
+		if err := clientStorage.WritePrompt(
+			ctx, core.UserInput{
```
Comment (diegohordi) [inline core/cmd/lambda/main.go:110]: **suggestion** for better readability, I would like to suggest creating a var to hold this core.UserInput value there.

File: core/cmd/lambda/main.go [hunk 5]
```diff
@@ -95,11 +104,40 @@ func handler(
...
+		}
+
 		graph, err := clientModel.Do(ctx, prompt)
 		if err != nil {
 			return corsHeaders.setHeaders(parseClientError(err)), err
 		}
 
+		prediction, _ := json.Marshal(graph)
+		if err := clientStorage.WriteModelPrediction(
+			ctx, core.ModelOutput{
```
Comment (diegohordi) [inline core/cmd/lambda/main.go:129]: suggestion for better readability, I would like to suggest creating a var to hold this core.ModelOutput value there.

File: core/cmd/lambda/main.go [hunk 6]
```diff
@@ -67,14 +76,14 @@ func (h corsHeaders) setHeaders(resp events.APIGatewayProxyResponse) events.APIG
 
 func handler(
```
Comment (diegohordi) [inline core/cmd/lambda/main.go:77]: **suggestion** To increase readability and decrease cognitive overload, I suggest refactoring this function and breaking it down into smaller ones.

File: core/storage/postgres.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+		v.Timestamp,
+	)
+	return err
+}
+
+const tableWriteModelPrediction = "public.openai_response"
+
+func (c client) WriteModelPrediction(ctx context.Context, v core.ModelOutput) error {
+	if v.UserID == "" {
+		return errors.New("user_id is required")
```
Comment (diegohordi) [inline core/storage/postgres.go:52]: **question** why not use sentinel errors to make these errors easier to check from the caller's perspective? https://dave.cheney.net/tag/errors

File: core/cmd/cli/main.go (hunk 1 above)
Comment (kislerdm) [reply core/cmd/cli/main.go:1]: Thanks for sharing the link! It's quite an interesting angle on the problem. IMO use of environment variables is meant to control runtime, the build [tags](https://pkg.go.dev/cmd/go#hdr-Build_constraints) is an officially recommended way to configure compiler.

File: core/cmd/lambda/handler_test.go (hunk 2 above)
Comment (kislerdm) [reply core/cmd/lambda/handler_test.go:15]: It's a good question, thanks for bringing it - it's quite educational. I feel like a tend to follow [recommendation](https://dmitri.shuralyov.com/idiomatic-go#use-singular-form-for-collection-repo-folder-name) to use `singular form` for package names. Exceptions: `utils`, `examples`.

Comment (kislerdm) [reply core/cmd/lambda/handler_test.go:15]: @diegohordi BTW there nothing "out of hands" in aliasing import, don't you feel? :)

File: core/cmd/lambda/main.go (hunk 3 above)
Comment (kislerdm) [reply core/cmd/lambda/main.go:86]: @diegohordi good proposal. What for though? Errors shall be treated, what treatment would you suggest for a connection failed to be closed?

File: core/cmd/lambda/main.go (hunk 4 above)
Comment (kislerdm) [reply core/cmd/lambda/main.go:110]: Thanks for your suggestion! I gladly apply it.

File: core/cmd/lambda/main.go (hunk 5 above)
Comment (kislerdm) [reply core/cmd/lambda/main.go:129]: Thanks for your suggestion! I gladly apply it.

File: core/cmd/lambda/main.go (hunk 6 above)
Comment (kislerdm) [reply core/cmd/lambda/main.go:77]: @diegohordi Good proposal. Please log an issue, we come back to it later if you don't mind. Thanks!

Comment (kislerdm) [reply core/cmd/lambda/main.go:77]: @diegohordi please do not hesitate to execute straight away if you feel like so - refactoring is always good when we don't notice invested effort ;)

File: core/storage/postgres.go (hunk 7 above)
Comment (kislerdm) [reply core/storage/postgres.go:52]: It is a good proposal. Let's log it and come back at refactoring.

For context: 
- the error is "private", end user will not see it;
- I intentionally didn't want to make commitment to a certain "error model" yet.

**Pull request merged**